lib/stream.js

```diff
@@ -90,43 +93,55 @@ const throwOnStreamError = async stream => {
 // The streams created by the `std*` options are automatically ended by `.pipe()`.
 // However `.pipe()` only does so when the source stream ended, not when it errored.
 // Therefore, when `childProcess.stdin|stdout|stderr` errors, those streams must be manually destroyed.
-const cleanupStdioStreams = (stdioStreams, error) => {
-	for (const stdioStream of stdioStreams) {
-		if (stdioStream.autoDestroy) {
-			stdioStream.value.destroy(error);
+const cleanupStdioStreams = (customStreams, error) => {
+	for (const customStream of customStreams) {
+		if (customStream.autoDestroy) {
+			customStream.value.destroy(error);
 		}
 	}
 };
 
 // Retrieve result of child process: exit code, signal, error, streams (stdout/stderr/all)
 export const getSpawnedResult = async (
 	spawned,
-	{encoding, buffer, maxBuffer},
+	{encoding, buffer, maxBuffer, timeout, killSignal, cleanup, detached},
+	context,
 	stdioStreams,
-	processDone,
 ) => {
+	const finalizers = [];
+	cleanupOnExit(spawned, cleanup, detached, finalizers);
+	const customStreams = getCustomStreams(stdioStreams);
+
 	const stdoutPromise = getStreamPromise(spawned.stdout, {encoding, buffer, maxBuffer});
 	const stderrPromise = getStreamPromise(spawned.stderr, {encoding, buffer, maxBuffer});
 	const allPromise = getStreamPromise(spawned.all, {encoding, buffer, maxBuffer: maxBuffer * 2});
 
 	try {
-		return await Promise.all([
-			processDone,
-			stdoutPromise,
-			stderrPromise,
-			allPromise,
-			throwIfStreamError(spawned.stdin, processDone),
-			...stdioStreams.map(stdioStream => waitForStreamEnd(stdioStream, processDone)),
+		return await Promise.race([
+			Promise.all([
+				once(spawned, 'exit'),
+				stdoutPromise,
+				stderrPromise,
+				allPromise,
+				...waitForCustomStreamsEnd(customStreams),
+			]),
+			...throwOnCustomStreamsError(customStreams),
+			...throwIfStreamError(spawned.stdin),
+			...throwOnTimeout({spawned, timeout, killSignal, context, finalizers}),
```

Comment:
`Promise.all()` waits for each argument and returns it: 
  - child process exit code and signal
  - stdout/stderr/all output
  - wait for any custom stream to complete, but do not use their output. For example, the stream created by using `{ stdout: './path.txt' }`.

`Promise.race()` short-circuits it by throwing if either:
  - custom streams error
  - `stdin` errors
  - on timeout, with the `timeout` option

The `throw*(...)` methods never resolve, they only throw.